Support to coercion of env vars to boolean

dlebrero opened the pull request.

weavejester commented:
Aren't booleans usually represented as vars that are either exist (truthy) or are missing/empty (falsey)?

dlebrero commented:
We wanted to have a var that was true by default. AFAIK, there is no way to make it empty. 

As a workaround, we could have inverted the name/meaning of the var, so instead of a var called "run-forever" with default value of true, we could have a var called "do-not-run-forever" with default value of empty/falsey, but we think this name makes the option confusing. Also, if anybody decides to specify the value as "false" it will unexpectedly be truthy, which will be very confusing.

weavejester commented:
I think this is reasonable, but I think the parser needs to be a little more accepting. My thought it that it should accept `true`, `false`, `t`, `f`, `yes`, `no`, `y` and `n` case insensitively, and the empty string as well as a synonym for false. Otherwise it should throw an exception.

weavejester commented on src/duct/core/env.clj:
```diff
@@ -12,6 +13,14 @@
+    (cond
+      (contains? #{"true" "t" "y" "yes"} x-lower-case) true
+      (contains? #{"false" "f" "n" "no" nil ""} x-lower-case) false
+      :else (throw (IllegalArgumentException.
```
Rather than `IllegalArgumentException`, perhaps using `ex-info` would be more Clojurey. Something like:

```clojure
(throw (ex-info (str "Could not coerce " (pr-str x) " into a boolean."
                     "Must be one of: \"true\", \"t\", \"false\", \"f\","
                     "\"yes\", \"y\", \"no\", \"n\", \"\" or nil.")
                {:value x, :coercion 'Bool}))
```

Also, can you keep lines to under 80 characters?

dlebrero replied:
Was using IllegalArgumentException to follow the 'Int coercion. Changed to ex-info

weavejester commented on test/duct/env_test.clj:
```diff
@@ -0,0 +1,27 @@
+    "no" 'Bool false
+    nil 'Bool false)
+
+
```
Remove double-spacing.

weavejester commented on test/duct/env_test.clj:
```diff
@@ -0,0 +1,27 @@
+            [duct.core.env :as env]))
+
+(deftest coerce
+  (are [a b c] (= (env/coerce a b) c)
```
Align clauses in `are`.

weavejester commented on test/duct/env_test.clj:
```diff
@@ -0,0 +1,27 @@
+
+  (is (thrown? IllegalArgumentException
+               (env/coerce "tru" 'Bool)))
+  )
```
Bracket shouldn't be on its own line.

weavejester commented on test/duct/env_test.clj:
```diff
@@ -0,0 +1,27 @@
+
+
+  (is (thrown? ArithmeticException (/ 1 0)))
+
```
Extra whitespace can be avoided without impacting visibility.

weavejester commented on test/duct/env_test.clj:
```diff
@@ -0,0 +1,27 @@
+    nil 'Bool false)
+
+
+  (is (thrown? ArithmeticException (/ 1 0)))
```
This test isn't testing anything in Duct. Accidentally included?

dlebrero replied:
Yes. Removed

weavejester commented on test/duct/env_test.clj:
```diff
@@ -0,0 +1,27 @@
+               (env/coerce "tru" 'Bool)))
+  )
+
+
```
Extra blank lines can be removed.

dlebrero commented on src/duct/core/env.clj:
```diff
@@ -12,6 +13,14 @@
 
+(defmethod coerce 'Bool [x _]
+  (let [x-lower-case (some-> x str/lower-case)]
+    (cond
```
Much better. Thanks.

dlebrero commented:
Thanks for the code review. 

I couldn't find any documentation in the Duct wiki regarding coercion. If you point me to the correct page I am happy to update it also.

weavejester commented on test/duct/env_test.clj:
```diff
@@ -0,0 +1,20 @@
+  (is (thrown? IllegalArgumentException
+               (env/coerce "abc" 'Int)))
+  (is (thrown? ExceptionInfo
+               (env/coerce "tru" 'Bool))))
```
Give it one newline at the end of the file. Tools tend to complain otherwise.

dlebrero replied:
Fixed

weavejester commented:
I don't think it's documented in the wiki, so feel free to add a page!

This looks good, except it needs an ending newline on the tests. Could you also squash your commits and change the message to something like:

    Add boolean coercion to #duct/env

sMy project adhere to the [seven rules](https://chris.beams.io/posts/git-commit/#seven-rules) for commit messages.

weavejester commented:
We can probably just copy the docstring into the wiki for now.

Unfortunately Github's "squash and merge" option is named somewhat misleadingly, as it doesn't generate a merge commit. Creating a clean commit history is still something that needs to be done through a manual rebase.

dlebrero commented:
Squashed and updated https://github.com/duct-framework/duct/wiki/Configuration#environment-variables 

Let me know if I am missing anything else.

Thanks!
